tests/serverTests/serverIndex.test.js

```diff
@@ -1,38 +1,149 @@
-    ));
-});
+  test('sever returns an array of reviews', (done) => {
+    const checkArray = (res) => {
+      if (!(Array.isArray(res.body))) {
+        done(new Error('not an array!'));
+      }
+    };
+
+    request(app)
+      .get('/restaurants/101/reviews')
+      .expect(checkArray)
+      .end(done);
+  });
+
+  // POST
+  test('server responds to successful POST requests with a 201', (done) => {
+    request(app)
+      .post('/restaurants/001/reviews')
+      .send({ rating: 5, review: 'I loved it!' })
+      .set('Content-Type', 'application/json')
+      .expect(201)
+      .end(err => (
+        err ? done(err) : done()
+      ));
+  });
+
+  test('POST requests create a new review', (done) => {
+    request(app)
+      .get('/restaurants/001/reviews')
+      .then((res) => {
+        [postedReview] = res.body;
+        if (res.body[0].reviewBody !== 'I loved it!') {
+          done(new Error('new review not created!'));
+        }
+        done();
+      });
+  });
+
+  test('server responds with 404 to POST requests without a review body', (done) => {
+    request(app)
+      .post('/restaurants/200/reviews')
+      .send({ rating: 5 })
+      .set('Content-Type', 'application/json')
+      .expect(404)
+      .end(err => (
+        err ? done(err) : done()
+      ));
+  });
+
+  test('server responds with 404 to POST requests without a review rating', (done) => {
+    request(app)
+      .post('/restaurants/200/reviews')
+      .send({ review: 'I loved it!' })
+      .set('Content-Type', 'application/json')
+      .expect(404)
+      .end(err => (
+        err ? done(err) : done()
+      ));
+  });
+
+  // PUT
+  test('server responds with 200 to a successful PUT request', (done) => {
+    const id = postedReview['_id']; // eslint-disable-line
 
-test('server responds to nonexistant routes with 404', (done) => {
-  request(app)
-    .get('/hello')
-    .expect(404)
-    .end(err => (
-      err ? done(err) : done()
-    ));
+    request(app)
+      .put(`/restaurants/001/reviews/${id}`)
+      .send({ cool: 9 })
+      .set('Content-Type', 'application/json')
+      .expect(200)
+      .end(err => (
+        err ? done(err) : done()
+      ));
+  });
+
+  test('Existing review is updated after a successful PUT request', (done) => {
+    request(app)
+      .get('/restaurants/001/reviews')
+      .expect((res) => {
+        if (res.body[0].cool !== 9) {
+          done(new Error('PUT request did not update successfully'));
+        }
+      })
+      .end(done);
+  });
+
+  test('server responds with 404 to PUT request without body', (done) => {
+    const id = postedReview['_id']; // eslint-disable-line
+
+    request(app)
+      .put(`/restaurants/001/reviews/${id}`)
+      .set('Content-Type', 'application/json')
+      .expect(404)
+      .end(err => (
+        err ? done(err) : done()
+      ));
+  });
+
+  test('sever responds with 404 to PUT requests with invalid properties to update', (done) => {
```

Comment:
Is there a reason you're handling the case of invalid property submissions on your server side? I'd think one would prefer handling these checks on the client side so that the client is not able to submit invalid information and a 404 response is not necessary

Reply:
You're right that I definitely want to do this check on the client side before the request is sent, but I think it's important to keep this check in place on server-side too. I don't want to rely on the server to send good requests.